### PR TITLE
Cascade terminate/purge support in GrpcDurableTaskClient

### PR DESCRIPTION
**Pull request by shivamkm07**
This PR adds support to recursively terminate/purge sub-orchestrations in GrpcDurableTaskClient. It also sets the recursive behavior to be true by default. 

Closes: #260

### CODE REVIEW

**Review by jviau — COMMENTED** [review]
As we now have multiple parameters for terminate, I recommend we do the following:

1. Add `TerminateInstanceOptions`
2. Make the current `abstract` method `virtual`. Default implementation will call one of the new methods we add.
3. Add `TerminateInstanceOptions` accepting `virtual` methods
4. Update our clients to override the new options-receiving method.

``` CSharp
/// add xmldoc
public record TerminateInstanceOptions
{
    public string? Output { get; init; }

    public bool Recursive { get; init; }
}
```

new methods:

``` CSharp
public virtual Task TerminateInstanceAsync(string instanceId, TerminateInstanceOptions? options, CancellationToken cancellation)
    => throw new NotSupportedException($"{this.GetType()} does not support orchestration termination.");

// this is the formerly abstract method
public virtual Task TerminateInstanceAsync(string instanceId, object? output = null, CancellationToken cancellation = default)
{
    TerminateInstanceOptions? options = output is null ? null : new() { Output = output };
    return this.TerminateInstanceAsync(instanceId, options, cancellation);
}
```

File: src/Client/Core/DurableTaskClient.cs [hunk 1]
```diff
@@ -238,13 +248,14 @@ public virtual Task TerminateInstanceAsync(string instanceId, CancellationToken
...
     /// <param name="output">The optional output to set for the terminated orchestration instance.</param>
+    /// <param name="recursive">The optional parameter to recursively termintate sub-orchestrations, true by default.</param>
     /// <param name="cancellation">
     /// The cancellation token. This only cancels enqueueing the termination request to the backend. Does not abort
     /// termination of the orchestration once enqueued.
     /// </param>
     /// <returns>A task that completes when the terminate message is enqueued.</returns>
     public abstract Task TerminateInstanceAsync(
-        string instanceId, object? output = null, CancellationToken cancellation = default);
+        string instanceId, object? output = null, bool recursive = true, CancellationToken cancellation = default);
```
[the remark this answers is not in the document]
Comment (shivamkm07) [reply src/Client/Core/DurableTaskClient.cs:258]: A separate virtual method is added, with same definition as previous abstract method. So anyone using this abstract method should work still now right? How is this a breaking change then?
 ```
public virtual Task TerminateInstanceAsync(string instanceId, object? output, CancellationToken cancellation)
        => this.TerminateInstanceAsync(instanceId, output, true, cancellation);

```

**Comment by shivamkm07** [comment]
> As we now have multiple parameters for terminate, I recommend we do the following:
> 
> 1. Add `TerminateInstanceOptions`
> 2. Make the current `abstract` method `virtual`. Default implementation will call one of the new methods we add.
> 3. Add `TerminateInstanceOptions` accepting `virtual` methods
> 4. Update our clients to override the new options-receiving method.
> 
> ```cs
> /// add xmldoc
> public record TerminateInstanceOptions
> {
>     public object? Output { get; init; }
> 
>     public bool Recursive { get; init; }
> }
> ```
> 
> new methods:
> 
> ```cs
> public virtual Task TerminateInstanceAsync(string instanceId, TerminateInstanceOptions? options, CancellationToken cancellation)
>     => throw new NotSupportedException($"{this.GetType()} does not support orchestration termination.");
> 
> // this is the formerly abstract method
> public virtual Task TerminateInstanceAsync(string instanceId, object? output = null, CancellationToken cancellation = default)
> {
>     TerminateInstanceOptions? options = output is null ? null : new() { Output = output };
>     return this.TerminateInstanceAsync(instanceId, options, cancellation);
> }
> ```

I am okay with this change. Just should not the new `TerminateInstanceOptions` method be an abstract method and both `GrpcDurableTaskClient` and `ShimDurableTaskClient` implement this method? This is what seems to be followed in other `DurableTaskClient` methods..

File: src/Client/Core/DurableTaskClient.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jviau) [reply src/Client/Core/DurableTaskClient.cs:258]: Correct on the `virtual` method meaning the old abstract method is covered. But now there is a _new_ abstract method which is a breaking change on its own.

File: src/Client/Core/DurableTaskClient.cs [hunk 2]
```diff
@@ -368,7 +392,7 @@ public virtual Task ResumeInstanceAsync(string instanceId, CancellationToken can
     /// <see cref="PurgeResult.PurgedInstanceCount"/> indicating the number of orchestration instances that were purged.
     /// </returns>
     public abstract Task<PurgeResult> PurgeAllInstancesAsync(
-        PurgeInstancesFilter filter, CancellationToken cancellation = default);
+        PurgeInstancesFilter filter, bool recursive = true, CancellationToken cancellation = default);
```
Comment (jviau) [inline src/Client/Core/DurableTaskClient.cs:395]: Same issue with the other comment: adding a net-new abstract method is a breaking change.

**Review by jviau — CHANGES_REQUESTED** [review]

File: src/Client/Core/DurableTaskClient.cs (hunk 2 above)
Comment (shivamkm07) [reply src/Client/Core/DurableTaskClient.cs]: Changed to virtual

File: src/Client/Core/DurableTaskClient.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (shivamkm07) [reply src/Client/Core/DurableTaskClient.cs]: Done

**Comment by shivamkm07** [comment]
> As we now have multiple parameters for terminate, I recommend we do the following:
> 
> 1. Add `TerminateInstanceOptions`
> 2. Make the current `abstract` method `virtual`. Default implementation will call one of the new methods we add.
> 3. Add `TerminateInstanceOptions` accepting `virtual` methods
> 4. Update our clients to override the new options-receiving method.
> 
> ```cs
> /// add xmldoc
> public record TerminateInstanceOptions
> {
>     public object? Output { get; init; }
> 
>     public bool Recursive { get; init; }
> }
> ```
> 
> new methods:
> 
> ```cs
> public virtual Task TerminateInstanceAsync(string instanceId, TerminateInstanceOptions? options, CancellationToken cancellation)
>     => throw new NotSupportedException($"{this.GetType()} does not support orchestration termination.");
> 
> // this is the formerly abstract method
> public virtual Task TerminateInstanceAsync(string instanceId, object? output = null, CancellationToken cancellation = default)
> {
>     TerminateInstanceOptions? options = output is null ? null : new() { Output = output };
>     return this.TerminateInstanceAsync(instanceId, options, cancellation);
> }
> ```

@jviau Pushed the suggested changes. Please review.

File: src/Client/Grpc/GrpcDurableTaskClient.cs [hunk 3]
```diff
@@ -124,19 +124,25 @@ public override ValueTask DisposeAsync()
 
     /// <inheritdoc/>
     public override async Task TerminateInstanceAsync(
-        string instanceId, object? output = null, CancellationToken cancellation = default)
+        string instanceId, TerminateInstanceOptions? options = null, CancellationToken cancellation = default)
     {
+        if (options is null)
+        {
+            options = new TerminateInstanceOptions();
```
Comment (jviau) [inline src/Client/Grpc/GrpcDurableTaskClient.cs:131]: I'd prefer we avoid allocating unnecessary objects like this. Instead you can either:

1. Have a `static` empty instance of `TerminateInstanceOptions`
2. OR split into defaults in this method:

``` CSharp
string? output = options?.Output;
bool recursive = options?.Recursive ?? false;

// rest of function
```

File: src/Client/OrchestrationServiceClientShim/ShimDurableTaskClient.cs [hunk 4]
```diff
@@ -169,11 +169,16 @@ public override AsyncPageable<OrchestrationMetadata> GetAllInstancesAsync(Orches
 
     /// <inheritdoc/>
     public override Task TerminateInstanceAsync(
-        string instanceId, object? output = null, CancellationToken cancellation = default)
+        string instanceId, TerminateInstanceOptions? options = null, CancellationToken cancellation = default)
     {
+        if (options is null)
+        {
+            options = new TerminateInstanceOptions();
```
Comment (jviau) [inline src/Client/OrchestrationServiceClientShim/ShimDurableTaskClient.cs:176]: Same here as the other comment: avoid unnecessary object allocation

**Review by jviau — CHANGES_REQUESTED** [review]

File: src/Client/Core/DurableTaskClient.cs [hunk 5]
```diff
@@ -353,22 +391,33 @@ public virtual Task ResumeInstanceAsync(string instanceId, CancellationToken can
...
     /// A <see cref="CancellationToken"/> that can be used to cancel the purge operation.
     /// </param>
     /// <returns>
     /// This method returns a <see cref="PurgeResult"/> object after the operation has completed with a
     /// <see cref="PurgeResult.PurgedInstanceCount"/> indicating the number of orchestration instances that were purged.
     /// </returns>
-    public abstract Task<PurgeResult> PurgeAllInstancesAsync(
-        PurgeInstancesFilter filter, CancellationToken cancellation = default);
+    public virtual Task<PurgeResult> PurgeAllInstancesAsync(
+        PurgeInstancesFilter filter, bool recursive = true, CancellationToken cancellation = default)
```
[the remark this answers is not in the document]
Comment (shivamkm07) [reply src/Client/Core/DurableTaskClient.cs]: Added `PurgeInstanceOptions` But haven't inherited from `PurgeInstancesFilter` because it is being used by both `PurgeInstanceAsync` and `PurgeAllInstancesAsync` while `PurgeInstancesFilter` is just used in `PurgeAllInstancesAsync`.

File: src/Client/Core/DurableTaskClient.cs [hunk 6]
```diff
@@ -338,13 +372,17 @@ public virtual Task ResumeInstanceAsync(string instanceId, CancellationToken can
...
     /// <see cref="OrchestrationRuntimeStatus.Terminated"/> state can be purged.
     /// </para><para>
+    /// Purging an orchestration will by default purge all of the child sub-orchestrations that were started by the
+    /// orchetration instance. If you don't want to purge sub-orchestration instances, you can set `recursive` flag to
+    /// false which will disable purging of child sub-orchestration instances.
     /// If <paramref name="instanceId"/> is not found in the data store, or if the instance is found but not in a
     /// terminal state, then the returned <see cref="PurgeResult"/> object will have a
     /// <see cref="PurgeResult.PurgedInstanceCount"/> value of <c>0</c>. Otherwise, the existing data will be purged and
-    /// <see cref="PurgeResult.PurgedInstanceCount"/> will be <c>1</c>.
+    /// <see cref="PurgeResult.PurgedInstanceCount"/> will be <c>countOfInstancesPurged</c>.
```
[the remark this answers is not in the document]
Comment (shivamkm07) [reply src/Client/Core/DurableTaskClient.cs]: Done

File: src/Client/Core/TerminateInstanceOptions.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.DurableTask.Client;
+
+/// <summary>
+/// Options to terminate an orchestration.
+/// </summary>
+public class TerminateInstanceOptions
```
[the remark this answers is not in the document]
Comment (shivamkm07) [reply src/Client/Core/TerminateInstanceOptions.cs]: Done

File: src/Client/Core/DurableTaskClient.cs [hunk 8]
```diff
@@ -209,12 +209,40 @@ protected DurableTaskClient(string name)
...
+
+    /// <inheritdoc cref="TerminateInstanceAsync(string, TerminateInstanceOptions, CancellationToken)"/>
+    public virtual Task TerminateInstanceAsync(string instanceId, bool recursive)
+    {
+        TerminateInstanceOptions options = new() { Recursive = recursive };
+        return this.TerminateInstanceAsync(instanceId, options);
+    }
+
+    /// <inheritdoc cref="TerminateInstanceAsync(string, TerminateInstanceOptions, CancellationToken)"/>
+    public virtual Task TerminateInstanceAsync(string instanceId, bool recursive, CancellationToken cancellation)
```
[the remark this answers is not in the document]
Comment (shivamkm07) [reply src/Client/Core/DurableTaskClient.cs]: Removed

File: src/Client/Core/DurableTaskClient.cs [hunk 9]
```diff
@@ -209,12 +209,40 @@ protected DurableTaskClient(string name)
     public abstract Task<OrchestrationMetadata> WaitForInstanceCompletionAsync(
         string instanceId, bool getInputsAndOutputs = false, CancellationToken cancellation = default);
 
-    /// <inheritdoc cref="TerminateInstanceAsync(string, object, CancellationToken)"/>
+    /// <inheritdoc cref="TerminateInstanceAsync(string, TerminateInstanceOptions, CancellationToken)"/>
     public virtual Task TerminateInstanceAsync(string instanceId, CancellationToken cancellation)
         => this.TerminateInstanceAsync(instanceId, null, cancellation);
 
+    /// <inheritdoc cref="TerminateInstanceAsync(string, TerminateInstanceOptions, CancellationToken)"/>
+    public virtual Task TerminateInstanceAsync(string instanceId, object? output)
```
[the remark this answers is not in the document]
Comment (shivamkm07) [reply src/Client/Core/DurableTaskClient.cs]: Done

File: src/Client/Grpc/GrpcDurableTaskClient.cs (hunk 3 above)
Comment (shivamkm07) [reply src/Client/Grpc/GrpcDurableTaskClient.cs]: Done

File: src/Client/OrchestrationServiceClientShim/ShimDurableTaskClient.cs (hunk 4 above)
Comment (shivamkm07) [reply src/Client/OrchestrationServiceClientShim/ShimDurableTaskClient.cs]: Done

File: src/Client/Core/DurableTaskClient.cs [hunk 10]
```diff
@@ -353,22 +391,33 @@ public virtual Task ResumeInstanceAsync(string instanceId, CancellationToken can
     /// <see cref="PurgeResult.PurgedInstanceCount"/> value of <c>1</c> or <c>0</c>, depending on whether the target
     /// instance was successfully purged.
     /// </returns>
-    public abstract Task<PurgeResult> PurgeInstanceAsync(
-        string instanceId, CancellationToken cancellation = default);
+    public virtual Task<PurgeResult> PurgeInstanceAsync(
+        string instanceId, bool recursive = true, CancellationToken cancellation = default)
```
[the remark this answers is not in the document]
Comment (shivamkm07) [reply src/Client/Core/DurableTaskClient.cs]: Added

File: src/Client/Core/DurableTaskClient.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jviau) [reply src/Client/Core/DurableTaskClient.cs]: Oh I meant the other way around:

`PurgeInstanceFilter : PurgeInstanceOptions`

File: src/Client/Core/PurgeInstanceOptions.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.DurableTask.Client;
+
+/// <summary>
+/// Options to purge an orchestration.
+/// </summary>
+public record PurgeInstanceOptions
```
Comment (jviau) [inline src/Client/Core/PurgeInstanceOptions.cs:9]: You can simplify a record type:

``` CSharp
public record PurgeInstanceOptions(bool Recursive = false);
```

Also, we need to keep recursive as default `false`. Otherwise, the default is a breaking behavior change from what we have today (which is non-recursive).

File: src/Client/Core/TerminateInstanceOptions.cs [hunk 12]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.DurableTask.Client;
+
+/// <summary>
+/// Options to terminate an orchestration.
+/// </summary>
+public record TerminateInstanceOptions
```
Comment (jviau) [inline src/Client/Core/TerminateInstanceOptions.cs:9]: Can simplify this type as well:

``` CSharp
public record TerminateInstanceOptions(object? Output = null, bool Recursive = false)
```

Also, we need to keep recursive as default `false`. Otherwise, the default is a breaking behavior change from what we have today (which is non-recursive).

File: src/Shared/Grpc/ProtoUtils.cs [hunk 13]
```diff
@@ -49,7 +49,6 @@ internal static HistoryEvent ConvertHistoryEvent(P.HistoryEvent proto)
                         OrchestrationInstance = proto.ExecutionStarted.ParentInstance.OrchestrationInstance.ToCore(),
                         TaskScheduleId = proto.ExecutionStarted.ParentInstance.TaskScheduledId,
                     },
-                    Correlation = proto.ExecutionStarted.CorrelationData,
```
Comment (jviau) [inline src/Shared/Grpc/ProtoUtils.cs:52]: Why is this removed? Was it removed from the proto?

File: src/Client/Grpc/GrpcDurableTaskClient.cs [hunk 14]
```diff
@@ -321,18 +325,20 @@ public override AsyncPageable<OrchestrationMetadata> GetAllInstancesAsync(Orches
 
     /// <inheritdoc/>
     public override Task<PurgeResult> PurgeInstanceAsync(
-        string instanceId, CancellationToken cancellation = default)
+        string instanceId, PurgeInstanceOptions? options = null, CancellationToken cancellation = default)
     {
+        bool recursive = options?.Recursive ?? true;
```
Comment (jviau) [inline src/Client/Grpc/GrpcDurableTaskClient.cs:330]: default to `false` to avoid breaking behavior change. Please also address this in all other locations for purge and terminate.

If we want the behavior to be `true` by default we will need to do that in a major version rev. You can add a TODO comment for that in the options class.

**Review by jviau — CHANGES_REQUESTED** [review]
Only a few small comments of recursive default behavior.

File: src/Client/Core/PurgeInstanceOptions.cs (hunk 11 above)
Comment (shivamkm07) [reply src/Client/Core/PurgeInstanceOptions.cs:9]: Actually default value is intentionally kept as `true` here, but I agree it would be a breaking change. The thing is it is already set as true at other places like `durabletask-python` and even in `Dapr`, so setting it to false here would make it inconsistent. thoughts here @cgillum ?

Comment (shivamkm07) [reply src/Client/Core/PurgeInstanceOptions.cs:9]: Okay, makes sense.

Comment (shivamkm07) [reply src/Client/Core/PurgeInstanceOptions.cs]: Done

File: src/Client/Core/TerminateInstanceOptions.cs (hunk 12 above)
Comment (shivamkm07) [reply src/Client/Core/TerminateInstanceOptions.cs]: Done

File: src/Shared/Grpc/ProtoUtils.cs (hunk 13 above)
Comment (shivamkm07) [reply src/Shared/Grpc/ProtoUtils.cs:52]: Yes. The build was failing once I updated the proto

File: src/Client/Grpc/GrpcDurableTaskClient.cs (hunk 14 above)
Comment (shivamkm07) [reply src/Client/Grpc/GrpcDurableTaskClient.cs]: Done

File: src/Client/OrchestrationServiceClientShim/ShimDurableTaskClient.cs
Comment (jviau) [inline src/Client/OrchestrationServiceClientShim/ShimDurableTaskClient.cs:1]: Please add a TODO comment to the purge and terminate functions here to support recursive.

**Review by jviau — APPROVED** [review]

File: src/Client/OrchestrationServiceClientShim/ShimDurableTaskClient.cs
Comment (shivamkm07) [reply src/Client/OrchestrationServiceClientShim/ShimDurableTaskClient.cs:1]: Done
